### DISCUSSION TITLE
不正な証明書に対して警告が頻繁に表示される

### DISCUSSION THREAD

**Discussion by sayurin** [ffftp/ffftp#358] [Ideas]
#9 により、不正な証明書に対して警告が表示されるようになったが、警告を抑止するオプションが提供されていないので、アプリケーションを起動し直す度に警告される。

- 起動中は一度許可すれば次からは警告されないので現状維持でよい。
- 保護されていないのであれば暗号化する意味はないので、利用者は不正な証明書を使用するサイトに対しftps接続せず、ftp接続すればよい。その場合は警告されることもなくなる。
- 設定項目の「暗号化なしで接続を許可」を「保護されていない接続を許可」に意味を変更する。この設定が有効な場合は不正な証明書に対して警告しない。

等の案があります。

**Comment by honeylover** [comment]
エックスサーバーの例で言うと、おそらく何年も前に契約した人は、FTPホスト名が ABCDEFG.xsrv.jp という形式で提供されていると思われます。先ほどエックスサーバーのコンパネで見たところ、FTPホスト名は sv**.xserver.jp という形式でした。

それで、後者の場合は証明書の *.xserver.jp に一致するので何のエラーも出ませんが、前者の形式の場合は一致しないためにエラー扱いになります。ところが、実際に接続されるサーバーはどちらの形式で書いたとしても当然ながら物理的に同じサーバーです。 

それでいて、片方は不正扱いにしかできないというのがちょっと納得しがたいところなのですが、Filezillaのように受け入れるかどうかをユーザーに問うというのは無理なのでしょうか？

ミニバードの方はちょっと違うケースかもしれませんが、ネットオウルが提供するサービスとしてのミニバードであるために証明書の方は *.minibird.netowl.jp なのだと思います（他のサービスでも共用なのかも）が、指定されたサーバー名は ****.minibird.jp であって一致しないものの、やはりユーザーが使用OKできるようになればいいんじゃないかと感じました。

他のケースとしては、FTPサーバーが独自ドメインで、証明書は実際の物理的なサーバー会社のものというケースがあります。当然一致しないという警告が出ますが、信頼に値するものであれば、そのまま暗号化して欲しいという思いがあります。

深いことは分からない素人なので無茶ぶりしているかもしれませんが、とにかくユーザーの自己責任において受け入れて暗号化通信ができればいいかなと思いました。

>保護されていないのであれば暗号化する意味はないので

#39 でも書きましたが、本当に【まったく】意味がなのかなと・・・。今回のようなケースなら意味が出てくるんじゃないかと思ったり。。。実は、Filezilla以外の他のソフトでも試してみましたが、一致しなくても暗号化しているものばかりなので、それはそれでありなんじゃないかなとか感じた次第です。

**Comment by sayurin** [comment]
> Filezillaのように受け入れるかどうかをユーザーに問うというのは無理なのでしょうか？

というのは **設定項目の「暗号化なしで接続を許可」を「保護されていない接続を許可」に意味を変更する。** が近いでしょうか？ もちろんホスト設定だけでなく、証明書に問題が見つかった際にも許可するかどうかを尋ねます。

> 本当に【まったく】意味がなのかなと・・・。今回のようなケースなら意味が出てくるんじゃないかと思ったり。。。実は、Filezilla以外の他のソフトでも試してみましたが、一致しなくても暗号化しているものばかりなので、それはそれでありなんじゃないかな

保護を目的に有償の証明書を購入しておいてそれが間違っていたというのは管理者の責任ですし、無意味な行為にコストをかけているのではサービスとしてどうかとは思います。また現在では無料で証明書を作成することもできるわけで、やはり管理者は正しく設定すべきことだとは思います。
有償・無償どちらであれ杜撰な管理をされたサーバーに対して、利用者が我慢する必要があるのか？ アプリケーションが対応する必要があるのか？ というのが私のスタンスでもあります。

とはいえ、私個人の考えですから、意見を求めたいというのが本Issueです。

**Comment by kawasumi2019** [comment]
横から失礼します。個人的な意見を書かせて頂きます。
意見としてはsayurinさんに非常に近く、証明書という1つのセキュリティを簡単に形骸化させることについては良いことではないと思っています。
現在の警告メッセージでは「保護されない接続を許可」という内容ですが、利用者の方を守るという観点では平文での接続さえも許可すべきでないと思います。

接続先と証明書の内容が合致していません。
接続先 => server1.example.com
証明書 => server2.example.jp
接続先は詐称されている可能性があります。
それでも接続しますか？　OR　ホストから切断しました。

これくらいの表現が必要だと思っています。
と同時に、気になる部分もあります。RFC4217によると、日本語での「保護」と同等の言葉と思われる「Protection」は暗号化の意味で使われています。誤解を避ける意味でsayurinさんの意図される「保護されていない接続」は別の言葉、例えば「未検証のホストへの接続」に換えられてもよいのではと考えます。

またFTPはプロトコル上どういう名前で接続されているかをサーバ側で知ることは出来ないため、異なるFQDNで同じIPアドレスを指す証明書を不正な証明書とサーバを運営されている側で認識していない可能性もあると思います。
正しい証明書を設定するべき（今回のお話であればドメインの追加）という前提はありますが、FTPサーバ側で出来ない検証をFFFTP（クライアント）側でやっていると考えれば、追加のオプションという形で「証明書による厳格なホスト検証をする」というものを増設されるのもまた一案ではないかと思いました。

**Comment by sayurin** [comment]
現在ある「暗号化なしで接続を許可」は非SSLホストを指します。案として挙げた「保護されていない接続」は非SSLホストに加えて誤った証明書のホストも指すイメージです。
また「保護」「Protection」についてはRFC・技術用語ではなく、Google Chromeでの表記・一般用語を意識しました。

どのようなホスト名で接続がなされているかわからないのはFTPに限らずSSLの問題です。これに対しSSLとしてはSNIを用意していますし、あまり関係ないかと思います。

**Comment by kawasumi2019** [comment]
返信ありがとうございます。
「保護されていない接続」はGoogle Chromeでの表記に合わせられているということで、疑問点が雲散霧消しました。なるほど、アドレスバーのあれだったのですね。
あれ自体をSSL接続ではないという意味に置き換えて読んでいたので、混乱してしまってすみません。
（誤った証明書の時も「保護されていません」と赤字で表示されますね。）

RFC7151が標準化されればSNI・・・と言うより「HOSTコマンド（小声）」「MUST（小声）」と囁いたらダッシュで対応してくれそうですけど、現段階だとSNIはHTTPSの技術という印象が強いです。
ヘッダにホスト名を持つHTTP/1.0と元々ホスト名を伝える手段のないFTPとではやはり差を感じますし、SSL通信を前提に考えれば逆に関係ない話になりますけれど、「FTPってそういうもの」と一度思ったら、そういうのはなかなか抜けないんじゃないかと思いました。
もちろん私自身は最初に書いた通り証明書の問題があれば接続自体をキャンセルしても良いんじゃないかと思っているので、その印象とstrictな判断をすべきかどうかは別と考えています。
